### PR TITLE
first use of Postgres DB to store user login credentials

### PR DESCRIPTION
**Pull request by joey-xu**
use Postgres user_id table to store (username,  hashed_password, salt, logged_in)

get rid of the internal map db_

### CODE REVIEW

File: server/db_accessor.cpp [hunk 1]
```diff
@@ -4,59 +4,125 @@
...
+    }
+    
+    bool logged_in = std::get<0>(logged_status.at(0));
+    if (!logged_in) {
+        BOOST_LOG_SEV(db_accessor::logger_, sev::warning) << "User " << id
+            << " is trying to logout when the status is already logged out.";
+        return false;
+    }
+    
+    ozo::execute(conn, "UPDATE user_id SET logged_in = false WHERE username="_SQL + id , yield[ec]);
```
Comment (three0s) [inline server/db_accessor.cpp]: I think we don't need to do any checking, just let the person log out... More sql = more opportunity to run into concurrency issues, especially without transactions and `select for update` etc.

File: server/db_accessor.h [hunk 2]
```diff
@@ -13,23 +21,31 @@ class db_accessor {
     struct user {
```
Comment (three0s) [inline server/db_accessor.h]: we don't need this anymore

File: server/db_accessor.cpp [hunk 3]
```diff
@@ -4,59 +4,125 @@
...
 
 constexpr std::size_t HASH_SIZE = 32;
+static auto conn_info = ozo::make_connection_info("host=127.0.0.1 port=5432 dbname=niba");
+nibaserver::logger db_accessor::logger_ = logger();
+
+bool nibaserver::db_accessor::login(boost::asio::io_context &ioc, boost::asio::yield_context &yield,
+    const std::string &id, const std::string &password) {
+    ozo::error_code ec{};
+    ozo::rows_of<ozo::pg::name, std::vector<char>, std::vector<char>, bool> user_credential;
+    auto conn = ozo::request(ozo::make_connector(conn_info, ioc),
```
Comment (three0s) [inline server/db_accessor.cpp]: I am thinking of passing in (or create one on server start up) a connection_pool or a  connection provider that the db_accessor has direct access to.
https://github.com/yandex/ozo/blob/master/examples/connection_pool.cpp
So the processor and session don't really need to know anything about `io_context`. And we only pass around the `yield_context`

Comment (three0s) [reply server/db_accessor.cpp]: thinking about it, we could have 1 connector (from a pool) per db_accessor, and 1 db_accessor per session, and give the processor a reference to the db_accessor instead

File: CMakeLists.txt [hunk 4]
```diff
@@ -39,6 +39,7 @@ target_link_libraries(niba-server
     Boost::log
     Boost::regex
     Threads::Threads
+    pq
```
[the remark this answers is not in the document]
Comment (three0s) [reply CMakeLists.txt]: let's make them submodules after this PR gets merged in

File: server/db_accessor.cpp (hunk 1 above)
Comment (joey-xu) [reply server/db_accessor.cpp]: Ok, now logout() does not check whether the username exists in the database or the current status is logged_in. There is only one UPDATE to set logged_in = false. This change saves one SELECT query.

File: server/db_accessor.cpp [hunk 5]
```diff
@@ -4,59 +4,125 @@
...
 }
 
-bool nibaserver::db_accessor::create_user(const std::string &id, const std::string &password) {
-    auto iter = db_.find(id);
-    if (iter != db_.end())
+bool nibaserver::db_accessor::create_user(boost::asio::io_context &ioc, boost::asio::yield_context &yield, 
+    const std::string &id, const std::string &password) {
+    ozo::error_code ec{};
+    ozo::rows_of<std::int64_t> username_count;
+    auto conn = ozo::request(ozo::make_connector(conn_info, ioc),
```
[the remark this answers is not in the document]
Comment (joey-xu) [reply server/db_accessor.cpp]: That is true.

**Comment by three0s** [comment]
Ship it!

File: server/db_accessor.cpp (hunk 3 above)
Comment (joey-xu) [reply server/db_accessor.cpp]: Done in the latest commit. Create the connection pool in main. Initialize the per-session db_accessor. Pass the connector into  db_accessor.

**Comment by three0s** [comment]
Can you also add the preprocessor definition (add it in to cmake like how we define client/server for CPRINT) and other ozo header file changes into our code?

File: server/main.cpp [hunk 6]
```diff
@@ -88,8 +108,9 @@ int main(int argc, char *argv[]) {
             tcp::socket socket(ioc);
             acceptor.async_accept(socket, yield);
             BOOST_LOG_SEV(logger, sev::info) << "Got connection";
+            nibaserver::db_accessor db(connector);
```
Comment (three0s) [inline server/main.cpp]: This db probably goes out of scope and gets destructed? Please double check

Or move it into the session(see how we do it for socket_ in server_session)

**Pull request merged**